### PR TITLE
Feature add exclude support for reposync

### PR DESCRIPTION
**Pull request by ghost**
For example, this package is broken in rhn:

```
gfs2-utils-3.0.12-23.el6.x86_6 FAILED
gfs2-utils-3.0.12-23.el6.x86_64: [Errno 256] No more mirrors to try.
WARNING:root:sync failed: rhel-6-x86_64-os

$ rm /var/repo/yum/rhel/6/x86_64/os/Packages/gfs2-utils-3.0.12-23.el6.x86_64.rpm
rm: remove regular empty file ‘/var/repo/yum/rhel/6/x86_64/os/Packages/gfs2-utils-3.0.12-23.el6.x86_64.rpm’?
```

Add exclude option:

```
$ cat /home/repo/config/rhel-6-x86_64.repo | grep exclu -B4 -A1
[rhel-6-x86_64-os]
auth = rhel-6-main
url = https://cdn.redhat.com/content/dist/rhel/server/6/6Server/x86_64/os
path = rhel/6/x86_64/os
exclude = gfs2-utils*
createrepo = true
```

Run with reposync:

```
...
Removing obsolete gfs2-utils-3.0.12-23.el6.x86_64.rpm
INFO:root:generating package metadata: rhel-6-x86_64-os
...
```

yum.conf(5)

```
exclude List of packages to exclude from all repositories, so yum works as if that package was never in  the
              repositories.  This  should be a space separated list.  This is commonly used so a package isn't upgraded or
              installed accidentally, but can be used to remove packages in any way that "yum list"  will  show  packages.
              Shell globs using wildcards (eg. * and ?) are allowed.
```

### CODE REVIEW

**Comment by pyther** [comment]
@mmyagkov Thanks for the pull request. I have some mixed feelings.

Regarding your particular issue, you should open a support case with Red Hat regarding the missing package. If this was a package such as udev or redhat-release you'd be in trouble! I have experienced this same problem. I have noticed that sometimes running a sync a few times will eventually fetch the package. As a matter of fact, I am currently dealing with this for an openstack repo and rhel-7-supplementary-sprm repo - I have engaged Red Hat.

Outside of this use case, I'm not sure when one would want to exclude packages from Red Hat repositories. I am actually rather surprised that reposync respected the exclude option in the yum config. Did you verify that the gfs2-utils package was not download? I've seen reposync fail to download a package the first time, but a second run results in successful download.

I lean towards not accepting this PR, as I see this as an unnecessary feature, but I could probably be convinced to easily change my mind.

File: upstream_sync.py [hunk 1]
```diff
@@ -24,10 +24,23 @@
...
+    :param mode: mode for makedirs
+    :param dir_path: dir path for check
+
+    :return: None
+    """
+    if not os.path.isdir(dir_path):
+        try:
+            os.makedirs(dir_path, mode)
+        except IOError as e:
+            print e
```
Comment (pyther) [inline upstream_sync.py:42]: We  eed to do a bit more than just print the error. If we can't create the directory there is no expectation the sync and createrepo commands will be successful. The quick solution is to print the error and then exit. The more elegant solution may be to handle the error and continue to the next repository.

**Comment by ghost** [comment]
@pyther Thanks for the feedback.

reposync exclude option may be required if you do not use KDE, Gnome, etc.. in production environment . Do not make sense to keep a copy of the full repository, if for understandable reasons, most will never be used.

Example:

```
$ find . -type f -name 'libreoffice*' | xargs du -sm | awk '{ sum+=$1} END {print sum "M"}'
4979M
$ find . -type f -name 'kde-*' | xargs du -sm | awk '{ sum+=$1} END {print sum "M"}'
2753M
$ find . -type f -name 'gnome-*' | xargs du -sm | awk '{ sum+=$1} END {print sum "M"}'
2026M
```

p.s. gfs2-utils now is ok.

```
$ curl --cacert /etc/rhsm/ca/redhat-uep.pem -E cert.pem --key key.pem https://cdn.redhat.com/content/dist/rhel/server/6/6Server/x86_64/os/Packages/gfs2-utils-3.0.12-23.el6.x86_64.rpm -o /dev/null
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  241k  100  241k    0     0  70245      0  0:00:03  0:00:03 --:--:-- 70231
```

File: upstream_sync.py (hunk 1 above)
Comment (ghost) [inline upstream_sync.py:42]: Yes, I totally agree. Hotfixed. More correctly I do a little bit later. Thanks!

**Comment by pyther** [comment]
@mmyagkov Thanks for the explanation. I live in a world where disk space is plentiful and I would prefer wasted storage than potentially broken repos, but I realize I'm lucky and can understand where you are coming from. I'll work on merging this in.

**Comment by ghost** [comment]
@pyther You say as a programmer, and not as a system administrator. I would prefer to work and updated repository here and now, rather than waiting on the RHN fixes. An exception can be removed completely when similar problems of this kind will be removed. Security issues are concerned not only there, where I came from ;)

File: upstream_sync.py [hunk 2]
```diff
@@ -24,10 +24,24 @@
...
+            sys.exit(1)
+
+
+def build_yum_config(name, url, sslcacert, sslcert, sslkey, exclude):
     # Check tmp path exist
-    if not os.path.isdir(tmp_dir):
-        os.mkdir(tmp_dir)
+    make_dir(tmp_dir)
```
Comment (pyther) [inline upstream_sync.py:48]: This changes the code behavior. In the previous code, the users umask would be used, but now directories will be created with 0700. Was this done intentionally, if so what is the reasoning? There shouldn't be any sensitive information in the generated repo file, so I think I would prefer to default to the users umask.

Comment (ghost) [inline upstream_sync.py:48]: You're right, I need to be a little easier when I see a \* .pem or \* .key in the configuration files. Current `umask` value should be a right decision.

**Pull request merged**